jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/QoSHandler.java

```diff
@@ -0,0 +1,465 @@
+            setMaxRequests(maxRequests);
+        }
+        state.set(maxRequests, 0);
+
+        if (LOG.isDebugEnabled())
+            LOG.debug("{} initialized maxRequests={}", this, maxRequests);
+
+        super.doStart();
+    }
+
+    @Override
+    protected void doStop() throws Exception
+    {
+        super.doStop();
+        removeBean(timeouts);
+        timeouts.destroy();
+    }
+
+    @Override
+    public boolean handle(Request request, Response response, Callback callback) throws Exception
+    {
+        if (LOG.isDebugEnabled())
+            LOG.debug("{} handling {}", this, request);
+
+        while (true)
+        {
+            long encoded = state.get();
+            int permits = AtomicBiInteger.getHi(encoded);
+            if (permits > 0)
+            {
+                if (!state.compareAndSet(encoded, permits - 1, 0))
+                    continue;
+                return forward(request, response, callback);
+            }
+            else
+            {
+                boolean expired = request.getAttribute(EXPIRED_ATTRIBUTE) != null;
+                if (expired)
+                {
+                    // Do not suspend again an attempt to handle the expired request/response.
+                    response.setStatus(HttpStatus.SERVICE_UNAVAILABLE_503);
+                    if (response.isCommitted())
+                        callback.failed(new IllegalStateException("Response already committed"));
+                    else
+                        response.write(true, null, callback);
+                }
+                else
+                {
+                    // Avoid this race condition:
+                    // T1 in handle() may find no permits, so it will suspend the request.
+                    // T2 in resume() finds no suspended requests and increments the permits.
+                    // T1 suspends the request, which will remain suspended despite permits are available.
+                    // Below, the suspended request count is only incremented if there are no permits.
+                    // See correspondent code in resume().
+                    int suspended = AtomicBiInteger.getLo(encoded);
+                    if (!state.compareAndSet(encoded, 0, suspended + 1))
+                        continue;
+                    suspend(request, response, callback);
+                }
+                return true;
+            }
+        }
+    }
+
+    /**
+     * <p>Returns the priority of the given suspended request,
+     * a value greater than or equal to {@code 0}.</p>
+     * <p>Priority {@code 0} is the lowest priority.</p>
+     * <p>The set of returned priorities should be stable over
+     * time, typically constrained in the range {@code 0-10}.</p>
+     *
+     * @param request the suspended request to compute the priority for
+     * @return the priority of the given suspended request, a value {@code >= 0}
+     */
+    protected int getPriority(Request request)
```

Comment:
There are no checks that `getPriority` won't return negative values. Also, why are those forbidden?